libass/ass_render.c

```diff
@@ -2140,6 +2141,50 @@ static void calculate_rotation_params(ASS_Renderer *render_priv, ASS_DRect *bbox
 }
 
 
+static int quantize_blur(double radius, int32_t *shadow_mask)
+{
+    // Gaussian filter kernel (1D):
+    // G(x, r2) = exp(-x^2 / (2 * r2)) / sqrt(2 * pi * r2),
+    // r = 64 * radius, r2 = r^2.
+
+    // Difference between kernels with different but near r2:
+    // G(x, r2 + dr2) - G(x, r2) ~= dr2 * G(x, r2) * (x^2 - r2) / (2 * r2^2).
+    // Maximal possible error relative to full pixel value
+    // is half of integral of absolute value of that difference.
+    // E_max ~= dr2 / 2 * integral(G(x, r2) * |x^2 - r2| / (2 * r2^2), x)
+    //  = dr2 / (4 * r2) * integral(G(y, 1) * |y^2 - 1|, y)
+    //  = dr2 / (4 * r2) * 4 / sqrt(2 * pi * e)
+    //  ~ dr2 / (4 * r2) ~= dr / (2 * r).
+    // E_max ~ BLUR_ACCURACY / 2 as we have 2 dimensions.
+
+    // To get discretized blur radius solve the following
+    // differential equation (n--quantization index):
+    // dr(n) / dn = BLUR_ACCURACY * r + TRANSFORM_ACCURACY, r(0) = 0,
```

Comment:
`TRANSFORM_ACCURACY` is acceptable error in spatial domain, i.&nbsp;e. we don't care about 1/8 pixel difference in blur radius even if it can lead to errors more than `BLUR_ACCURACY` in final result. It's mostly needed for regularization at small blur radius.

`n` is simply integer number and `r(n)` is quantized radius number `n`. I'm substituting discrete `r` with continuous function for simplicity so `dr(n) / dn` is simply `dr` from previous formula.

Reply:
> Could you explain in more detail why `TRANSFORM_ACCURACY` is needed? Isnt’t it desirable that we allow errors only below `BLUR_ACCURACY` in the final result?

If we allow 1/8 pixel transform accuracy then we allow 256/8=32 error near sharp edges. Well, solution of `dr / dn = BLUR_ACCURACY * r` is singular near zero radius so some regularization is needed. And it make sense to use `r ~= TRANSFORM_ACCURACY * n` for small `r`.

Question about error in final result is always tricky: it's always possible to stack enough layers to make final error arbitrary large. As for error of single blur--there is dithering in my blur code that reduces banding but adds error, so actual maximum error will be around 1.0+1.0 but it's only for worst pixels in worst cases. I think it looked good enough for naked eye, but we can always tweak that value or even add API quality switch (but I think we should switch to 10-bit color or higher first).